zebra-consensus/src/parameters/network_upgrade.rs

```diff
@@ -8,12 +8,18 @@ use std::ops::Bound::*;
 use zebra_chain::types::BlockHeight;
 use zebra_chain::{Network, Network::*};
 
-/// A Zcash network protocol upgrade.
-//
-// TODO: are new network upgrades a breaking change, or should we make this
-//       enum non-exhaustive?
+/// A Zcash network upgrade.
+///
+/// Network upgrades can change the Zcash network protocol or consensus rules in
+/// incompatible ways.
 #[derive(Copy, Clone, Debug, Eq, Hash, PartialEq)]
 pub enum NetworkUpgrade {
+    /// The Zcash protocol for a Genesis block.
+    ///
+    /// Zcash genesis blocks use a different set of consensus rules from
+    /// other BeforeOverwinter blocks, so we treat them like a separate network
+    /// upgrade.
```

Comment:
I'm skeptical of the complexity trade-off here. There's exactly one rule that differs for the genesis bug, which is just a bug: the height-in-coinbase was missing.

Reply:
Hmm, good point. I'm not sure how we'll use the `NetworkUpgrades` enum yet.

But depending on the outcome of the parallel verification RFC in #763, we might have some additional Zebra special cases:
* there are no VerificationConstraints, because the genesis block is the first block (except, perhaps, that the genesis block is the first block to be committed to the state)
* There is no previous verification context
* the on-disk verification state is empty
* the verification context and on-disk state need to be initialised with empty values
* the verification context needs to be updated with the values from the genesis block, based on genesis-specific rules
* there are no chain tips in memory or on disk - the genesis block automatically becomes the first tip

See https://github.com/ZcashFoundation/zebra/blob/330f9b16154543cf17fd8a785b29a70b26b733da/book/src/designs/0002-parallel-verification.md#chain-context-and-network-upgrades

I think a lot of these special cases are covered by checkpoint verification, but we'll definitely need special handling for the on-disk state.

I'll review the need for the `Genesis` network upgrade after the RFC.

Reply:
I think it still makes sense to have a distinct `Genesis` variant, because (unfortunately) the rules are actually different. Downstream code for whom this distinction doesn't matter can combine arms of a match statement, but not making the distinction at all means that we can't handle the differences in the rules in a principled way.

Reply:
Just wanted to make sure you'd thought about the complexity tradeoffs — sounds like you have.